src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -211,304 +219,28 @@ private static void EscapeNextBytes(byte value, Span<byte> destination, ref int
-
             Debug.Assert(encoder != null);
 
-            // Convert char to byte.
-            byte[] utf8DestinationArray = null;
-            Span<byte> utf8Destination;
-            int length = checked((value.Length) * JsonConstants.MaxExpansionFactorWhileTranscoding);
-            if (length > JsonConstants.StackallocThreshold)
-            {
-                utf8DestinationArray = ArrayPool<byte>.Shared.Rent(length);
-                utf8Destination = utf8DestinationArray;
-            }
-            else
-            {
-                unsafe
-                {
-                    byte* ptr = stackalloc byte[JsonConstants.StackallocThreshold];
-                    utf8Destination = new Span<byte>(ptr, JsonConstants.StackallocThreshold);
-                }
-            }
-
-            ReadOnlySpan<byte> utf16Value = MemoryMarshal.AsBytes(value);
-            OperationStatus toUtf8Status = ToUtf8(utf16Value, utf8Destination, out int bytesConsumed, out int bytesWritten);
-
-            Debug.Assert(toUtf8Status != OperationStatus.DestinationTooSmall);
-            Debug.Assert(toUtf8Status != OperationStatus.NeedMoreData);
-
-            if (toUtf8Status != OperationStatus.Done)
-            {
-                if (utf8DestinationArray != null)
-                {
-                    utf8Destination.Slice(0, bytesWritten).Clear();
-                    ArrayPool<byte>.Shared.Return(utf8DestinationArray);
-                }
+            OperationStatus result = encoder.Encode(value, destination, out int encoderBytesConsumed, out int encoderCharsWritten);
 
-                ThrowHelper.ThrowArgumentException_InvalidUTF8(utf16Value.Slice(bytesWritten));
-            }
-
-            Debug.Assert(toUtf8Status == OperationStatus.Done);
-            Debug.Assert(bytesConsumed == utf16Value.Length);
-
-            // Escape the bytes.
-            byte[] utf8ConvertedDestinationArray = null;
-            Span<byte> utf8ConvertedDestination;
-            length = checked(bytesWritten * JsonConstants.MaxExpansionFactorWhileEscaping);
-            if (length > JsonConstants.StackallocThreshold)
-            {
-                utf8ConvertedDestinationArray = ArrayPool<byte>.Shared.Rent(length);
-                utf8ConvertedDestination = utf8ConvertedDestinationArray;
-            }
-            else
-            {
-                unsafe
-                {
-                    byte* ptr = stackalloc byte[JsonConstants.StackallocThreshold];
-                    utf8ConvertedDestination = new Span<byte>(ptr, JsonConstants.StackallocThreshold);
-                }
-            }
-
-            EscapeString(utf8Destination.Slice(0, bytesWritten), utf8ConvertedDestination, indexOfFirstByteToEscape: 0, encoder, out int convertedBytesWritten);
+            Debug.Assert(result != OperationStatus.DestinationTooSmall);
+            Debug.Assert(result != OperationStatus.NeedMoreData);
+            Debug.Assert(encoderBytesConsumed == value.Length);
```

Comment:
Would this be true if `OperationStatus != Done` but rather InvalidData? Looks like we should add a test to cover that scenario.

Reply:
This look like an invalid Assert. The same for Utf8. I'll see if I can add a test for Utf8; utf16 currently will not return InvalidData